loader/loader.c

```diff
@@ -6860,359 +6939,382 @@ VkResult setup_loader_term_phys_dev_groups(struct loader_instance *inst) {
-                           idx);
-                total_count = idx;
-                res = VK_ERROR_OUT_OF_HOST_MEMORY;
-                goto out;
+                new_phys_dev_groups[idx] = (VkPhysicalDeviceGroupPropertiesKHR *)loader_instance_heap_alloc(
+                    inst, sizeof(VkPhysicalDeviceGroupPropertiesKHR), VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+                if (NULL == new_phys_dev_groups[idx]) {
+                    loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
+                               "setup_loader_term_phys_dev_groups:  Failed to allocate physical device group Terminator object %d",
+                               idx);
+                    total_count = idx;
+                    res = VK_ERROR_OUT_OF_HOST_MEMORY;
+                    goto out;
+                }
+                memcpy(new_phys_dev_groups[idx], &local_phys_dev_groups[new_idx].group_props,
+                       sizeof(VkPhysicalDeviceGroupPropertiesKHR));
             }
-            memcpy(new_phys_dev_groups[idx], &local_phys_dev_groups[new_idx].group_props,
-                   sizeof(VkPhysicalDeviceGroupPropertiesKHR));
-        }
 
-        ++idx;
+            ++idx;
+        }
     }
 
 out:
 
-    if (VK_SUCCESS != res) {
-        if (NULL != new_phys_dev_groups) {
-            for (uint32_t i = 0; i < total_count; i++) {
-                loader_instance_heap_free(inst, new_phys_dev_groups[i]);
+    if (NULL != pPhysicalDeviceGroupProperties) {
+        if (VK_SUCCESS != res) {
+            if (NULL != new_phys_dev_groups) {
+                for (uint32_t i = 0; i < total_count; i++) {
+                    loader_instance_heap_free(inst, new_phys_dev_groups[i]);
+                }
+                loader_instance_heap_free(inst, new_phys_dev_groups);
             }
-            loader_instance_heap_free(inst, new_phys_dev_groups);
-        }
-    } else {
-        // Free everything that didn't carry over to the new array of
-        // physical device groups
-        if (NULL != inst->phys_dev_groups_term) {
-            for (uint32_t i = 0; i < inst->phys_dev_group_count_term; i++) {
-                bool found = false;
-                for (uint32_t j = 0; j < total_count; j++) {
-                    if (inst->phys_dev_groups_term[i] == new_phys_dev_groups[j]) {
-                        found = true;
-                        break;
+        } else {
+            // Free everything that didn't carry over to the new array of
+            // physical device groups
+            if (NULL != inst->phys_dev_groups_term) {
+                for (uint32_t i = 0; i < inst->phys_dev_group_count_term; i++) {
+                    bool found = false;
+                    for (uint32_t j = 0; j < total_count; j++) {
+                        if (inst->phys_dev_groups_term[i] == new_phys_dev_groups[j]) {
+                            found = true;
+                            break;
+                        }
+                    }
```

Comment:
Here we search the new list before deciding whether to delete it, while in `setup_loader_term_phys_devs` we do not search, just check for NULL. Is there a reason for the inconsistency?

Reply:
These all should be consistent now.  The problem was, the mechanism that was used for the NULL checks could fail if we encountered an out of memory scenario.  So, I went back and made them all consistent.  It's more work on the success case, but it will make sure things work properly in all scenarios.